File: backend/MiniBilge.Application/DTOs/Badge/BadgeDto.cs
namespace MiniBilge.Application.DTOs.Badge;

public class BadgeDto
{
    public Guid Id { get; set; }
    public string Key { get; set; } = string.Empty;
    public string Name { get; set; } = string.Empty;
    public string Description { get; set; } = string.Empty;
    public string Emoji { get; set; } = string.Empty;
    public string Category { get; set; } = string.Empty;
    public string Rarity { get; set; } = string.Empty;
    public bool IsEarned { get; set; }
    public DateTime? EarnedAt { get; set; }

    /// <summary>Bu rozetin, sorgulanan profil türü (child/adult) için geçerli olup olmadığı.</summary>
    public bool IsApplicableToProfile { get; set; } = true;

    /// <summary>Kilitli rozetler için sayısal ilerleme; hesaplanamıyorsa null.</summary>
    public BadgeProgressDto? Progress { get; set; }
}

/// <summary>Kilitli bir rozetin hedefe doğru ilerlemesi (örn. 3/5, %80).</summary>
public class BadgeProgressDto
{
    public int Current { get; set; }
    public int Target { get; set; }

    /// <summary>"count" | "categories" | "streak" | "days" | "percent"</summary>
    public string Unit { get; set; } = "count";
}

public class BadgeCollectionDto
{
    public int TotalBadges { get; set; }
    public int EarnedCount { get; set; }
    public List<BadgeDto> Badges { get; set; } = new();
}

File: mobile/minibilge_mobile/lib/features/collection/models/badge_dto.dart
class BadgeDto {
  final String id;
  final String key;
  final String name;
  final String description;
  final String emoji;
  final String category;
  final String rarity;
  final bool isEarned;
  final DateTime? earnedAt;
  final bool isApplicableToProfile;
  final BadgeProgressDto? progress;

  const BadgeDto({
    required this.id,
    required this.key,
    required this.name,
    required this.description,
    required this.emoji,
    required this.category,
    required this.rarity,
    required this.isEarned,
    this.earnedAt,
    this.isApplicableToProfile = true,
    this.progress,
  });

  factory BadgeDto.fromJson(Map<String, dynamic> json) => BadgeDto(
        id: json['Id'] as String,
        key: json['Key'] as String,
        name: json['Name'] as String,
        description: json['Description'] as String,
        emoji: json['Emoji'] as String,
        category: json['Category'] as String,
        rarity: json['Rarity'] as String,
        isEarned: json['IsEarned'] as bool? ?? false,
        earnedAt: json['EarnedAt'] != null
            ? DateTime.tryParse(json['EarnedAt'] as String)
            : null,
        isApplicableToProfile: json['IsApplicableToProfile'] as bool? ?? true,
        progress: json['Progress'] != null
            ? BadgeProgressDto.fromJson(json['Progress'] as Map<String, dynamic>)
            : null,
      );
}

class BadgeProgressDto {
  final int current;
  final int target;
  final String unit; // "count" | "categories" | "streak" | "days" | "percent"

  const BadgeProgressDto({
    required this.current,
    required this.target,
    required this.unit,
  });

  double get ratio => target <= 0 ? 0 : (current / target).clamp(0, 1).toDouble();

  factory BadgeProgressDto.fromJson(Map<String, dynamic> json) => BadgeProgressDto(
        current: json['Current'] as int? ?? 0,
        target: json['Target'] as int? ?? 0,
        unit: json['Unit'] as String? ?? 'count',
      );
}

class BadgeCollectionDto {
  final int totalBadges;
  final int earnedCount;
  final List<BadgeDto> badges;

  const BadgeCollectionDto({
    required this.totalBadges,
    required this.earnedCount,
    required this.badges,
  });

  factory BadgeCollectionDto.fromJson(Map<String, dynamic> json) =>
      BadgeCollectionDto(
        totalBadges: json['TotalBadges'] as int,
        earnedCount: json['EarnedCount'] as int,
        badges: (json['Badges'] as List)
            .map((e) => BadgeDto.fromJson(e as Map<String, dynamic>))
            .toList(),
      );
}
